test/e2e_node/image_list.go

```diff
@@ -151,27 +152,39 @@ func PrePullAllImages() error {
 	}
 	images := framework.ImageWhiteList.List()
 	klog.V(4).Infof("Pre-pulling images with %s %+v", puller.Name(), images)
+
+	var wg sync.WaitGroup
+	var once sync.Once
+	wg.Add(len(images))
 	for _, image := range images {
-		var (
-			err    error
-			output []byte
-		)
-		for i := 0; i < maxImagePullRetries; i++ {
-			if i > 0 {
-				time.Sleep(imagePullRetryDelay)
+		go func(image string) {
+			defer wg.Done()
+
+			var (
+				pullErr error
+				output  []byte
+			)
+			for i := 0; i < maxImagePullRetries; i++ {
+				if i > 0 {
+					time.Sleep(imagePullRetryDelay)
+				}
+				if output, pullErr = puller.Pull(image); pullErr == nil {
+					break
+				}
+				klog.Warningf("Failed to pull %s as user %q, retrying in %s (%d of %d): %v",
+					image, usr.Username, imagePullRetryDelay.String(), i+1, maxImagePullRetries, pullErr)
 			}
-			if output, err = puller.Pull(image); err == nil {
-				break
+			if pullErr != nil {
+				klog.Warningf("Could not pre-pull image %s %v output: %s", image, pullErr, output)
+				once.Do(func() {
+					err = pullErr
```

Comment:
Using `once.Do` is definitely one option here. I'm wondering if you considered any alternatives? For example, we could also concatenate all of the errors from pulling individual images.

Reply:
Yes, We definitely can use **[]error** to save all errors happened and concatenate it，But i didn't think it's meaningful, if the Caller just compare the **err** with **nil**.

Reply:
I don't have strong opinions :) Maybe a slight preference for saving as much error information as possible, but I'm flexible :)